onnx/onnx_backend.h

```diff
@@ -0,0 +1,914 @@
+#ifdef ONNX_BUILD_LIBRARY
+#define ONNX_PUBLIC __declspec(dllexport)
+#else
+#define ONNX_PUBLIC __declspec(dllimport)
+#endif
+#else
+#define ONNX_PUBLIC
+#endif
+#endif
+
+#if defined(ONNX_BACKEND_SUFFIX)
+#define ONNX_SYMBOL_CONCAT_(prefix, suffix) prefix##suffix
+#define ONNX_SYMBOL_CONCAT(prefix, suffix) ONNX_SYMBOL_CONCAT_(prefix, suffix)
+#define ONNX_SYMBOL_NAME(symbol_name) ONNX_SYMBOL_CONCAT(symbol_name, ONNX_BACKEND_SUFFIX)
+#else
+#define ONNX_SYMBOL_NAME(symbol_name) symbol_name
+#endif
+
+#include <stddef.h>
+
+#if !defined(ONNX_NO_STDINT_H)
+#if defined(_MSC_VER) && (_MSC_VER < 1600)
+typedef signed __int8 int8_t;
+typedef unsigned __int8 uint8_t;
+typedef signed __int16 int16_t;
+typedef unsigned __int16 uint16_t;
+typedef signed __int32 int32_t;
+typedef unsigned __int32 uint32_t;
+typedef signed __int64 int64_t;
+typedef unsigned __int64 uint64_t;
+#else
+#include <stdint.h>
+#endif
+#endif /* !defined(ONNX_NO_STDINT_H) */
+
+/**
+ * Opaque ONNX backend handle.
+ * ONNX backend is a combination of software layer and hardware device used to
+ * run an ONNX graph.
+ */
+typedef void* onnxBackend;
+/** Opaque ONNX graph handle. */
+typedef void* onnxGraph;
+
+/** Return code for ONNX functions */
+typedef int32_t onnxStatus;
+/**
+ * Type for enumeration values.
+ *
+ * The low 32 bits are reserved for standardized ONNX values.
+ * The high 32 bits are reserved for vendor-specific extensions. Applications
+ * must check for specific vendor extensions before interpreting these bits.
+ */
+typedef uint64_t onnxEnum;
+/**
+ * Type for bit fields.
+ *
+ * The low 32 bits are reserved for standardized ONNX values.
+ * The high 32 bits are reserved for vendor-specific extensions. Applications
+ * must check for specific vendor extensions before interpreting these bits.
+ */
+typedef uint64_t onnxBitfield;
+/**
+ * Type for pointers or handles for memory buffers.
+ * This type is intended to work not only for CPU-addressable memory, but also
+ * for device memory. uint64_t ensures the API can accomodate Vulkan buffers.
+ */
+typedef uint64_t onnxPointer;
+
+#define ONNX_STATUS_SUCCESS 0x0000
+#define ONNX_STATUS_FALLBACK 0x0001
+#define ONNX_STATUS_INVALID_INDEX 0x0010
+#define ONNX_STATUS_INVALID_SIZE 0x0011
+#define ONNX_STATUS_INVALID_POINTER 0x0012
+#define ONNX_STATUS_INVALID_PROTOBUF 0x0013
+#define ONNX_STATUS_INVALID_MODEL 0x0014
+#define ONNX_STATUS_INVALID_BACKEND 0x0015
+#define ONNX_STATUS_INVALID_GRAPH 0x0016
+#define ONNX_STATUS_INVALID_NAME 0x0017
+#define ONNX_STATUS_INVALID_SHAPE 0x0018
+#define ONNX_STATUS_UNSUPPORTED_VERSION 0x0020
+#define ONNX_STATUS_UNSUPPORTED_OPERATOR 0x0021
+#define ONNX_STATUS_UNSUPPORTED_PARAMETER 0x0022
+#define ONNX_STATUS_UNIDENTIFIED_NAME 0x0023
+#define ONNX_STATUS_NO_SYSTEM_MEMORY 0x0030
+#define ONNX_STATUS_NO_DEVICE_MEMORY 0x0031
+#define ONNX_STATUS_NO_SYSTEM_RESOURCES 0x0032
+#define ONNX_STATUS_NO_DEVICE_RESOURCES 0x0033
+#define ONNX_STATUS_INTERNAL_ERROR 0x0034
```

Comment:
Any reason not to use enums for these (and all the other enumerations in this header)?

Reply:
enums do not have a portable ABI, they can be 8-bit wide on some platforms/compilers/compiler flags (e.g. `-fshort-enums` in gcc).